File: MVP.Sample/MVP.Base/BaseUserControl/BaseUserControl.cs
using System;
using System.Collections.Generic;
using System.Web.UI;
using MVP.Base.BasePresenter;

namespace MVP.Base.BaseUserControl
{
    public class BaseUserControl : UserControl, IBaseUserControl
    {
        #region Properties
        public IBasePresenter Presenter { get; set; }
        //public virtual string ControlKey { get { return GetType().FullName; } }
        public virtual List<string> ChildControls { get; set; } //Store the fullname of (Children)Usercontrols' Type 
        public virtual string ParentControl { get; set; }

        public virtual List<string> RelatedControls { get; set; }

        #endregion

        #region Public Methods
        public virtual void AttachEventHandler() { }

        public void ShowErrorMessage(string message)
        {
            if (Page.Master == null || !(Page.Master is IMasterPage)) return;
            var masterPage = (IMasterPage)Page.Master;
            masterPage.ShowErrorMessage(message);
        }

        public string GetResource(string className, string resourceKey)
        {
            var result = GetGlobalResourceObject(className, resourceKey);
            return result != null ? result.ToString() : string.Empty;
        }

        public string GetResource(string resourceKey)
        {
            return GetResource(GetDefautResourceClassName(), resourceKey);
        }

        public virtual string GetDefautResourceClassName()
        {
            return "Common";
        }

        public virtual void ReLoad (){}

        public void CommitChanges()
        {
            if (!string.IsNullOrEmpty(ParentControl) && RelatedControls != null && RelatedControls.Count > 0)
            {
                if (BaseControlCollection.Controls.ContainsKey(ParentControl))
                {
                    var parentControl = BaseControlCollection.Controls[ParentControl];
                    if (parentControl != null)
                    {
                        foreach (var controlKey in RelatedControls)
                        {
                            parentControl[controlKey].ReLoad();
                        }
                    }
                }
            }
        }

        #endregion

        #region Event Handlers

        protected override void OnLoad(EventArgs e)
        {
            base.OnLoad(e);
            BaseControlCollection.Register(ParentControl, ID, this);
            AttachEventHandler();
        }

        #endregion
    }
}

File: DotNetNuke/DNN.MVP.Base/BaseUserControl/BaseUserControl.vb
Imports DotNetNuke.Entities.Modules.Actions
Imports DNN.MVP.Base.BasePresenter
Imports System.Web.UI
Imports DotNetNuke.Entities.Modules

Namespace BaseUserControl
    Public Class BaseUserControl(Of TPresenter As {Class, IBasePresenter})
        Inherits PortalModuleBase
        Implements IBaseUserControl, IActionable

#Region "Properties"
        Public Property Presenter() As TPresenter
        Public Overridable Property ParentControl As String
        Public Overridable Property RelatedControls As List(Of String)
#End Region

#Region "Public Methods"
        Public Overridable Sub AttachEventHandler()

        End Sub

        Public Function GetResource(ByVal className As String, ByVal resourceKey As String) As String
            Dim result = GetGlobalResourceObject(className, resourceKey)
            Return If(result IsNot Nothing, result.ToString(), String.Empty)
        End Function

        Public Function GetResource(ByVal resourceKey As String) As String
            Return GetResource(GetDefautResourceClassName(), resourceKey)
        End Function

        Public Overridable Function GetDefautResourceClassName() As String
            Return "Common"
        End Function
#End Region

#Region "IBaseUserControl"
        Public Overridable Sub ReLoad() Implements IBaseUserControl.ReLoad

        End Sub

        Public Sub CommitChanges() Implements IBaseUserControl.CommitChanges
            If (Not String.IsNullOrEmpty(ParentControl) AndAlso RelatedControls IsNot Nothing AndAlso RelatedControls.Count > 0) Then
                If (BaseControlCollection.Controls.ContainsKey(ParentControl)) Then
                    Dim parentCtr = BaseControlCollection.Controls(ParentControl)
                    If parentCtr IsNot Nothing Then
                        For Each controlKey As String In RelatedControls
                            parentCtr(controlKey).ReLoad()
                        Next
                    End If
                End If
            End If
        End Sub
#End Region

#Region "Event Handlers"
        Protected Sub Page_Init(ByVal sender As Object, ByVal e As EventArgs) Handles MyBase.Init
            Presenter = DirectCast(Activator.CreateInstance(GetType(TPresenter), Me), TPresenter)
        End Sub

        Protected Sub Page_Load(ByVal sender As Object, ByVal e As EventArgs) Handles MyBase.Load
            Register(ParentControl, ID, Me)
            AttachEventHandler()
        End Sub
#End Region

#Region "IActionable"
        Public Overridable ReadOnly Property ModuleActions() As ModuleActionCollection Implements IActionable.ModuleActions
            Get
                Return Nothing
            End Get
        End Property
#End Region
        
    End Class
End Namespace

